### PR TITLE
feat: add search text-sql pair in cached table

### PR DESCRIPTION
**Pull request by jitingxu1**
Related issue: #46 

Search question to check if it is in the cached table. 
- pass a cached `Table`
  - schema: question, sql, dialect
- create a search-text-task
  - search the cached table by calculating the [levenshtein normalized edit distance](https://rapidfuzz.github.io/Levenshtein/levenshtein.html#ratio)
- search question before generating sql by LLM

--------------------
**ADDED**

```
from ibis_birdbrain.bot import Bot
import ibis
ibis.options.interactive = True

con = ibis.duckdb.connect()
t = ibis.read_parquet("m.parquet")
con.create_table("mortgage", t.to_pyarrow())


questions = [
    {
        "question": "how many rows in the table",
        "sql": "select count(*) from mortgage",
        "dialect": "duckdb",
    },
    {
        "question": "top 10 rows in the table",
        "sql": "select * from mortgage limit 10",
        "dialect": "duckdb",
    },
]
cached_table = ibis.memtable(questions)
bot = Bot(con=con, cached_table=cached_table)

bot("top 10 rows in the table")
```

### CODE REVIEW

File: src/ibis_birdbrain/bot.py [hunk 1]
```diff
@@ -89,12 +90,20 @@ def __init__(
         body = """TODO"""  # noqa
 
         # system initialization message
+        attachments = [
+                DatabaseAttachment(con, description=self.data_description),
+        ]
+        # need a better way to handle this
+        if str(cached_table) != "None":
```
Comment (drabastomek) [inline src/ibis_birdbrain/bot.py]: Nit: Do we need to convert it to str? Unless the `cached_table` can return `"None"` as a string -- we should be just able to do `if cached_table is not None`, right?

File: src/ibis_birdbrain/flows/data.py [hunk 2]
```diff
@@ -42,35 +48,32 @@ def __call__(self, messages: Messages) -> Messages:
...
-        data_description = database_attachment.description
-
-        table_attachments = []
-        for attachment in messages[-1].attachments:
-            table_attachments.append(messages[-1].attachments[attachment])
+        database_attachment = messages[0].attachments.get_attachment_by_type(DatabaseAttachment)
+        table_attachments = messages[-1].attachments.get_attachment_by_type(TableAttachment)  
 
+        # check if question is found in cached table
+        search_task_response = self.tasks["search-cahced-question"](messages)
```
Comment (drabastomek) [inline src/ibis_birdbrain/flows/data.py]: Typo: should be `search-cached-question` I think.

File: src/ibis_birdbrain/tasks/sql.py [hunk 3]
```diff
@@ -14,7 +16,57 @@
...
+@ibis.udf.scalar.python
+def levenshtein_ratio(text1: str, text2: str) -> float:
+    return ratio(text1, text2)
+
 # tasks
+class SearchTextTask(Task):
+    """Ibis Birdbrain task to search cached text and SQL pair."""
+
+    def __init__(
+        self, name: str = "search-cahced-question", description: str = "Ibis Birdbrain search task"
```
Comment (drabastomek) [inline src/ibis_birdbrain/tasks/sql.py]: Typo here too.

**Review by drabastomek — COMMENTED** [review]
Some small things. The rest LGTM.

File: src/ibis_birdbrain/bot.py (hunk 1 above)
Comment (jitingxu1) [reply src/ibis_birdbrain/bot.py]: You are right. I was using `Basebackend`, it did not work for `None`. Thanks a lot.

**Comment by lostmygithubaccount** [comment]
merging, thanks!

**Pull request merged**